// @colyseus/schema/lib/types/TypeContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeContext = void 0;
const Metadata_1 = require("../Metadata");
const Schema_1 = require("../Schema");
const symbols_1 = require("./symbols");
class TypeContext {
    /**
     * For inheritance support
     * Keeps track of which classes extends which. (parent -> children)
     */
    static { this.inheritedTypes = new Map(); }
    static register(target) {
        const parent = Object.getPrototypeOf(target);
        if (parent !== Schema_1.Schema) {
            let inherits = TypeContext.inheritedTypes.get(parent);
            if (!inherits) {
                inherits = new Set();
                TypeContext.inheritedTypes.set(parent, inherits);
            }
            inherits.add(target);
        }
    }
    constructor(rootClass) {
        this.types = {};
        this.schemas = new Map();
        this.hasFilters = false;
        this.parentFiltered = {};
        if (rootClass) {
            this.discoverTypes(rootClass);
        }
    }
    has(schema) {
        return this.schemas.has(schema);
    }
    get(typeid) {
        return this.types[typeid];
    }
    add(schema, typeid = this.schemas.size) {
        // skip if already registered
        if (this.schemas.has(schema)) {
            return false;
        }
        this.types[typeid] = schema;
        //
        // Workaround to allow using an empty Schema (with no `@type()` fields)
        //
        if (schema[Symbol.metadata] === undefined) {
            Metadata_1.Metadata.init(schema);
        }
        this.schemas.set(schema, typeid);
        return true;
    }
    getTypeId(klass) {
        return this.schemas.get(klass);
    }
    discoverTypes(klass, parentIndex, parentFieldViewTag) {
        if (!this.add(klass)) {
            return;
        }
        // add classes inherited from this base class
        TypeContext.inheritedTypes.get(klass)?.forEach((child) => {
            this.discoverTypes(child, parentIndex, parentFieldViewTag);
        });
        const metadata = (klass[Symbol.metadata] ??= {});
        // if any schema/field has filters, mark "context" as having filters.
        if (metadata[symbols_1.$viewFieldIndexes]) {
            this.hasFilters = true;
        }
        if (parentFieldViewTag !== undefined) {
            this.parentFiltered[`${this.schemas.get(klass)}-${parentIndex}`] = true;
        }
        for (const fieldIndex in metadata) {
            const index = fieldIndex;
            const fieldType = metadata[index].type;
            const viewTag = metadata[index].tag;
            if (typeof (fieldType) === "string") {
                continue;
            }
            if (Array.isArray(fieldType)) {
                const type = fieldType[0];
                // skip primitive types
                if (type === "string") {
                    continue;
                }
                this.discoverTypes(type, index, viewTag);
            }
            else if (typeof (fieldType) === "function") {
                this.discoverTypes(fieldType, viewTag);
            }
            else {
                const type = Object.values(fieldType)[0];
                // skip primitive types
                if (typeof (type) === "string") {
                    continue;
                }
                this.discoverTypes(type, index, viewTag);
            }
        }
    }
}
exports.TypeContext = TypeContext;
